api/errors/errors.go

```diff
@@ -0,0 +1,318 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package errors
+
+import (
+	"bytes"
+	"fmt"
+	"strings"
+
+	"go.uber.org/yarpc/api/errors/codes"
+)
+
+// TODO: What to do with Unknown? Should we use use this for Application?
+// TODO: What to do with Internal? Should we have a function to create an error of this code?
+// TODO: should we have specific fields for each error code instead of keyValues ...string?
+
+// Code will extract the Code from a yarpc error.
+//
+// This will return None if the given error is nil or not a yarpc error.
+// This function is the defined way to test if an error is a yarpc error.
+func Code(err error) codes.Code {
+	if err == nil {
+		return codes.None
+	}
+	yarpcError, ok := err.(*yarpcError)
+	if !ok {
+		return codes.None
+	}
+	return yarpcError.Code
+}
+
+// Name will extract the user-defined error name from a yarpc error.
+//
+// This will return empty is the given error is nil, not a yarpc error, or the
+// Code is not Application.
+func Name(err error) string {
+	if err == nil {
+		return ""
+	}
+	yarpcError, ok := err.(*yarpcError)
+	if !ok {
+		return ""
+	}
+	if yarpcError.Code != codes.Application {
+		return ""
+	}
+	return yarpcError.Name
+}
+
+// Details will extract a copy the error details from a yarpc error.
+//
+// This will return empty if the given error is nil or not a yarpc error.
+func Details(err error) map[string]string {
+	if err == nil {
+		return nil
+	}
+	yarpcError, ok := err.(*yarpcError)
+	if !ok {
+		return nil
+	}
+	return copyStringStringMap(yarpcError.Details)
+}
+
+// WithCode converts the given error to have the given code.
+//
+// If the error is nil, this will return nil.
+// If the error is a yarpc error, this will convert the error, as well as add the original error code as a cause.
+// If the error is not a yarpc error, this will create a new error with the given code.
+// Note that the codes None and Application should not be used, if they are, the original error will be returned.
+// TODO: what to do if None or Application are used?
+func WithCode(err error, code codes.Code, keyValues ...string) error {
+	if err == nil {
+		return nil
+	}
+	switch code {
+	case codes.None, codes.Application:
+		return err
+	}
+	yarpcError, ok := err.(*yarpcError)
+	if !ok {
+		return newWellKnownYarpcError(code, append([]string{"error", err.Error()}, keyValues...))
+	}
+	if yarpcError.Code == code {
+		return yarpcError
+	}
+	c := yarpcError.copyAndAdd(keyValues...)
+
+	i := 1
+	for {
+		if _, ok := c.Details[fmt.Sprintf("cause-%d", i)]; !ok {
+			break
+		}
+		i++
+	}
+	c.add(fmt.Sprintf("cause-%d", i), c.Code.String())
+
+	c.Code = code
+	return c
+}
```

Comment:
We’ve reached an agreement that `WithCode` is unnecessary. The fmt-like constructors should be sufficient.